crawler.go

```diff
@@ -329,21 +359,65 @@ func (c *Crawler) scanRequestWork(workCh chan chan *http.Request, closeCh chan i
 			spider.reqch <- requestAndChan{req: req, ch: resc}
 			select {
 			case re := <-resc:
-				closeRequest(req)
 				if re.err != nil {
 					c.logf("crawler: send HTTP request got error: %v", re.err)
 				} else {
-					go func(res *http.Response) {
+					go func(clonedReq *http.Request, res *http.Response) {
 						defer closeResponse(res)
 						defer func() {
 							if r := recover(); r != nil {
 								c.logf("crawler: Handler got panic error: %v", r)
 							}
 						}()
+
+						var retryResponseCode bool
```

Comment:
May make sense to factor this out into a helper method to make it easier to read. So this reads as:

```go
if !c.shouldRetry(resp) {
  closeRequest(req)
  h, _ := c.Handler(res)
  h.ServeSpider(c.writeCh, res)
  return
}
// Retry the request
...
```